doc/sonic-application-extention/sonic-versioning-strategy.md

```diff
@@ -0,0 +1,117 @@
+
+## SONiC Package Releases
+
+As of today (202012 release), all SONiC Docker Images are released together with SONiC OS. With SONiC Application Extension
+Infrastructure SONiC Dockers are released independently. On every release an increment in package version number is required.
+The exact number in version that needs to be incremented depends on the type of changes comparing to previously released
+package.
+
+Types of changes:
+  - Breaking change: backwards incompatible change; reflects in ***major*** version number
+  - New functionality: backward compatible change; reflects in ***minor*** version number
+  - Bug fixes or enhancement; reflects in ***patch*** version number
+
+## Conventional Commits
+
+In order to help ***package maintainers*** to understand whether a breaking change was introduced
+comparing to previous release or a new functionality was included all SONiC repositories *can*
+follow conventional commits ():
+
+e.g:
+
+```
+feat: Introduce new methods in ConsumerTable
+
+BREAKING CHANGE: this feature breaks the Consumer/Producer based IPC
+```
+
+## SONiC Packages Versioning
+
+- Package release ***happens on SONiC release/branch-out*** or on demand for single Docker image: ***sub-release***
+- Manual version update ***is required*** when SONiC Package releases
+- Within a release major and minor version ***must*** not change
+  - Package API backward compatibility is promised
+  - No new functionality is included
+- On package release ***package maintainer must*** check package API compatibility with previous release
+- In case API changed comparing to previous release ***package maintainer must*** must increment major version in master prior to branch-out
+- In case API didn't change comparing to previous release ***package maintainer must*** must increment minor version in master prior branch-out
+- Minor version in master after branch-out ***must*** be incremented by ***package maintainer*** in order to avoid version overlap between branches
+- *Patch* version is updated on bug fixes and minor enhancements. Since this is a tedious work to do mannually on every change
+  this can be done only when the package is published to users via container registry (package release).
+  A package is published with a bug fix or enhancement whenever ***package maintainer decides*** to do so.
+- On package release ***package maintainer must*** update package dependencies
+
+<p align=center>
+<img src="img/versioning-strategy.svg" alt="Figure 1. SONiC Docker Images Versioning Strategy">
+</p>
+
+***package maintainer can*** update *default-reference* in package.json in SONiC buildimage to point to a default version which will be used when user installs a package.
+
+## Base OS versioning
+
+## Base OS API that a package uses
+
+- SONiC utilities; Until CLI autogeneration, sonic utilities plugin system API incompatibility must be recorded by base OS major version increment
```

Comment:
I think we should de-link the CLI autogeneration from this.  If we think the sonic-utilities is part of the base-os, we should list here. Alternately, we can let other individual dockers to register their dependencies on sonic-utilities.

CLI is just one form of management, and even for CLI there are multiple models (linux-like,  cisco-like etc). The base-os should not have any awareness of the CLI, or any UI.

So, I suggest we look at  whether we need sonic-util as part of base-os, independent of a specific CLI model